internal/controller/garbage_collector_controller.go

```diff
@@ -0,0 +1,170 @@
+	garbageCollectionReconcileTime        = defaultGarbageCollectionReconcileTime
+)
+
+// GarbageCollectIgnoreAnnotation garbage collector will skip NodeMaintenance with this annotation.
+const GarbageCollectIgnoreAnnotation = "maintenance.nvidia.com/garbage-collector.ignore"
+
+// NewGarbageCollectorOptions creates new *GarbageCollectorOptions
+func NewGarbageCollectorOptions() *GarbageCollectorOptions {
+	return &GarbageCollectorOptions{
+		pendingMaxNodeMaintenanceTime: defaultMaxNodeMaintenanceTime,
+		maxNodeMaintenanceTime:        defaultMaxNodeMaintenanceTime,
+	}
+}
+
+// GarbageCollectorOptions are options for GarbageCollector where values
+// are stored by external entity and read by GarbageCollector.
+type GarbageCollectorOptions struct {
+	sync.Mutex
+
+	pendingMaxNodeMaintenanceTime time.Duration
+	maxNodeMaintenanceTime        time.Duration
+}
+
+// Store maxNodeMaintenanceTime
+func (gco *GarbageCollectorOptions) Store(maxNodeMaintenanceTime time.Duration) {
+	gco.Lock()
+	defer gco.Unlock()
+
+	gco.pendingMaxNodeMaintenanceTime = maxNodeMaintenanceTime
+}
+
+// Load loads the last Stored options
+func (gco *GarbageCollectorOptions) Load() {
+	gco.Lock()
+	defer gco.Unlock()
+
+	gco.maxNodeMaintenanceTime = gco.pendingMaxNodeMaintenanceTime
+}
+
+// MaxNodeMaintenanceTime returns the last loaded MaxUnavailable option
+func (gco *GarbageCollectorOptions) MaxNodeMaintenanceTime() time.Duration {
+	return gco.maxNodeMaintenanceTime
+}
+
+// NewNodeMaintenanceGarbageCollector creates a new NodeMaintenanceGarbageCollector
+func NewNodeMaintenanceGarbageCollector(kClient client.Client, options *GarbageCollectorOptions, log logr.Logger) *NodeMaintenanceGarbageCollector {
+	return &NodeMaintenanceGarbageCollector{
+		Client:  kClient,
+		options: options,
+		log:     log,
+	}
+}
+
+// NodeMaintenanceGarbageCollector performs garbage collection for NodeMaintennace
+type NodeMaintenanceGarbageCollector struct {
+	client.Client
+
+	options *GarbageCollectorOptions
+	log     logr.Logger
+}
+
+// SetupWithManager sets up NodeMaintenanceGarbageCollector with controller manager
+func (r *NodeMaintenanceGarbageCollector) SetupWithManager(mgr ctrl.Manager) error {
+	return mgr.Add(r)
+}
+
+// Reconcile collects garabage once
+func (r *NodeMaintenanceGarbageCollector) Reconcile(ctx context.Context) error {
+	r.log.Info("periodic reconcile start")
+	r.options.Load()
+	r.log.V(log.DebugLevel).Info("loaded options", "maxNodeMaintenanceTime", r.options.MaxNodeMaintenanceTime())
+
+	mnl := &maintenancev1.NodeMaintenanceList{}
+	err := r.List(ctx, mnl)
+	if err != nil {
+		return errors.Wrap(err, "failed to list NodeMaintenance")
+	}
+
+	timeNow := time.Now()
+	for _, nm := range mnl.Items {
+		// skip NodeMaintenance with
+		nmLog := r.log.WithValues("namespace", nm.Namespace, "name", nm.Name)
+
+		if metav1.HasAnnotation(nm.ObjectMeta, GarbageCollectIgnoreAnnotation) && nm.Annotations[GarbageCollectIgnoreAnnotation] == "true" {
```

Comment:
nit: you can remove the HasAnnotation function call here and below. You will get an empty string if you try to read a non-existent key from map[string]string

Reply:
done.